kernel/aster-nix/src/sched/priority_scheduler.rs

```diff
@@ -17,40 +16,118 @@ pub fn init() {
+        f(local_rq);
+    }
+}
+
+struct PreemptRunQueue<T: PreemptSchedEntity> {
+    current: Option<Arc<T>>,
     /// Tasks with a priority of less than 100 are regarded as real-time tasks.
-    real_time_tasks: SpinLock<LinkedList<TaskAdapter>>,
+    real_time_tasks: VecDeque<Arc<T>>,
     /// Tasks with a priority greater than or equal to 100 are regarded as normal tasks.
-    normal_tasks: SpinLock<LinkedList<TaskAdapter>>,
+    normal_tasks: VecDeque<Arc<T>>,
+    should_preempt: bool,
 }
 
-impl PreemptScheduler {
+impl<T: PreemptSchedEntity> PreemptRunQueue<T> {
     pub fn new() -> Self {
         Self {
-            real_time_tasks: SpinLock::new(LinkedList::new(TaskAdapter::new())),
-            normal_tasks: SpinLock::new(LinkedList::new(TaskAdapter::new())),
+            current: None,
+            real_time_tasks: VecDeque::new(),
+            normal_tasks: VecDeque::new(),
+            should_preempt: false,
         }
     }
 }
 
-impl Scheduler for PreemptScheduler {
-    fn enqueue(&self, task: Arc<Task>) {
-        if task.is_real_time() {
-            self.real_time_tasks.lock_irq_disabled().push_back(task);
-        } else {
-            self.normal_tasks.lock_irq_disabled().push_back(task);
+impl<T: Sync + Send + PreemptSchedEntity> LocalRunQueue<T> for PreemptRunQueue<T> {
+    fn update_current(&mut self, flags: UpdateFlags) -> bool {
+        match flags {
+            UpdateFlags::Tick => {
+                if let Some(current) = &self.current {
+                    !current.is_real_time() && !self.real_time_tasks.is_empty()
+                } else {
+                    true
+                }
+            }
+            _ => true,
         }
     }
 
-    fn dequeue(&self) -> Option<Arc<Task>> {
-        if !self.real_time_tasks.lock_irq_disabled().is_empty() {
-            self.real_time_tasks.lock_irq_disabled().pop_front()
+    fn dequeue_current(&mut self) -> Option<Arc<T>> {
+        self.current.take()
+    }
+
+    fn pick_next_current(&mut self) -> Option<Arc<T>> {
+        if !self.real_time_tasks.is_empty() {
+            self.real_time_tasks.pop_front()
         } else {
-            self.normal_tasks.lock_irq_disabled().pop_front()
+            self.normal_tasks.pop_front()
+        }
+    }
+
+    fn set_current(&mut self, next: Option<Arc<T>>) {
+        let prev = self.current.take();
+        self.current = next;
+        if let Some(prev_task) = prev {
+            if prev_task.is_real_time() {
+                self.real_time_tasks.push_back(prev_task)
+            } else {
+                self.normal_tasks.push_back(prev_task);
+            }
         }
     }
 
-    fn should_preempt(&self, task: &Arc<Task>) -> bool {
-        !task.is_real_time() && !self.real_time_tasks.lock_irq_disabled().is_empty()
+    fn current(&self) -> Option<&Arc<T>> {
+        self.current.as_ref()
+    }
+
+    fn set_should_preempt(&mut self, should_preempt: bool) {
+        self.should_preempt = should_preempt
+    }
```

Comment:
> I do believe the decision of whether to set this flag should be made by ostd instead of scheduler developers since it's **policy-independent**.

No, it's policy-dependent at least on Linux. The task in the `stop` scheduling class cannot be preempted.

Reply:
Sorry I didn't put it clear. 

Let's take `update_current` as an example. Its main task is updating scheduling statistics of the current task (or maybe the runqueue). Its return value indicates whether we need a reschedule. How the return value is computed is **policy-dependent**.

We `update_current` in many cases, e.g. yielding, sleeping, or just ticking. We simply don't want to update `should_preempt` flag according to the return value in the case of yielding and sleeping. That is, whether we should use the return value is dependent on the use case, which is **policy-independent**.

So of course you can always return a false value from `update_current` when implementing, say stop scheduling policy, to avoid being preempted. This part is still under the control of policy makers.

Please let me know if I made anything wrong.

Reply:
I think it reasonable to turn `should_preempt` cpu-local. This way we can achieve cheaper access without lock mechanisms or disabled irq and therefore a fast-path in `schedule`.

```rust
cpu_local! {
    static SHOULD_PREEMPT: AtomicBool = AtomicBool::new(false);
}

/// Invokes a schedule.
pub fn schedule() {
    if !SHOULD_PREEMPT.load(Ordering::Relaxed) {
        return;
    }
    SHOULD_PREEMPT.store(false, Ordering::Relaxed);

    reschedule(&mut |local_rq| {

        if let Some(next_current) = local_rq.pick_next_current() {
            local_rq.set_current(Some(next_current.clone()));
            ReschedAction::SwitchTo(next_current)
        } else {
            ReschedAction::DoNothing
        }
    });
}
```